Update handling of negative radius for feMorphology

Ahmad-S792 opened the pull request:
#### e7825a0cbca5b51b5fd45c8be7a9a07802cffa98
<pre>
Update handling of negative radius for feMorphology

<a href="https://bugs.webkit.org/show_bug.cgi?id=257838">https://bugs.webkit.org/show_bug.cgi?id=257838</a>

Reviewed by Simon Fraser.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: <a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=202644

Spec[1] states:

&quot;A negative or zero value disables the effect of the given filter
  primitive (i.e., the result is the filter input image).&quot;

So this changes FEMorphology to clamp radii to non-negative and stop treating
as an error in SVGFEMorphologyElement::createFilterEffect.

[1] <a href="https://drafts.fxtf.org/filters/#element-attrdef-femorphology-radius">https://drafts.fxtf.org/filters/#element-attrdef-femorphology-radius</a>

* Source/WebCore/platform/graphics/filters/FEMorphology.cpp:
(FEMorphology::FEMorphology): clamp &apos;m_radiusX&apos; and &apos;m_radiusY&apos;
(FEMorphology::setRadiusX): clamp &apos;radiusX&apos;
(FEMorphology::setRadiusY): clamp &apos;radiusY&apos;
* Source/WebCore/svg/SVGFEMorphologyElement.cpp:
(SVGFEMorphologyElement::isIdentity): clamp return as per commit
(SVGFEMorphologyElement::createFilterEffect): Remove condition to return &apos;nullptr&apos; on negative radius
* LayoutTests/svg/filters/feMorphology-negative-radius.html: Add Test Case
* LayoutTests/svg/filters/feMorphology-negative-radius-expected.html: Add Test Case Expectation
* LayoutTests/svg/filters/feMorphology-radius-cases.svg: Rebaselined

Canonical link: <a href="https://commits.webkit.org/265007@main">https://commits.webkit.org/265007@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f744d96a58804a182c974002de386db32efbe0a0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9443 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9715 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9950 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11105 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/9300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9451 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11700 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9675 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12189 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9593 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10497 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11262 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7791 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8607 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16036 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8888 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8756 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12090 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9257 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7532 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8462 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/8513 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2280 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/12687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9020 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa reviewed (commented):
I think we need to change `SVGFEMorphologyElement::isIdentity()` as well. Negative or zero radius should be considered as a no-op effect:

```
bool SVGFEMorphologyElement::isIdentity() const
{
    return radiusX() <= 0 && radiusY() <= 0;
}
```

or 

```
bool SVGFEMorphologyElement::isIdentity() const
{
    return !std::max(0f, radiusX()) && !std::max(0f, radiusY());
}
```

Ahmad-S792 committed:
Update handling of negative radius for feMorphology

https://bugs.webkit.org/show_bug.cgi?id=257838

Reviewed by Simon Fraser.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: https://src.chromium.org/viewvc/blink?view=revision&revision=202644

Spec[1] states:

"A negative or zero value disables the effect of the given filter
  primitive (i.e., the result is the filter input image)."

So this changes FEMorphology to clamp radii to non-negative and stop treating
as an error in SVGFEMorphologyElement::createFilterEffect.

[1] https://drafts.fxtf.org/filters/#element-attrdef-femorphology-radius

* Source/WebCore/platform/graphics/filters/FEMorphology.cpp:
(FEMorphology::FEMorphology): clamp 'm_radiusX' and 'm_radiusY'
(FEMorphology::setRadiusX): clamp 'radiusX'
(FEMorphology::setRadiusY): clamp 'radiusY'
* Source/WebCore/svg/SVGFEMorphologyElement.cpp:
(SVGFEMorphologyElement::isIdentity): clamp return as per commit
(SVGFEMorphologyElement::createFilterEffect): Remove condition to return 'nullptr' on negative radius
* LayoutTests/svg/filters/feMorphology-negative-radius.html: Add Test Case
* LayoutTests/svg/filters/feMorphology-negative-radius-expected.html: Add Test Case Expectation
* LayoutTests/svg/filters/feMorphology-radius-cases.svg: Rebaselined

Canonical link: https://commits.webkit.org/265007@main
